clippy_lints/src/and_then_then_some.rs

```diff
@@ -0,0 +1,164 @@
+    pub AND_THEN_THEN_SOME,
+    nursery,
+    "detects usage of and_then and then_some that can be replaced by filter"
+}
+
+// note: `Option::filter` is older than `bool::then_some`,
+// so no msrv check is required.
+declare_lint_pass!(AndThenThenSome => [AND_THEN_THEN_SOME]);
+
+impl<'tcx> LateLintPass<'tcx> for AndThenThenSome {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        match expr.kind {
+            ExprKind::MethodCall(_, selfarg, [arg], _) | ExprKind::Call(_, [selfarg, arg]) => {
+                // TODO: check if type of reciever is diagnostic item Option?
+                if is_and_then(cx, expr) {
+                    if let Some((closure_args, predicate)) = then_some_closure_arg(cx, arg) {
+                        show_sugg(cx, expr.span, selfarg, closure_args, predicate);
+                    }
+                }
+            },
+            _ => {},
+        }
+    }
+}
+
+// `|v| X.then_some(v)` -> Some((span"|v|", X))
+fn then_some_closure_arg<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) -> Option<(Span, &'tcx Expr<'tcx>)> {
+    match expr.kind {
+        ExprKind::Closure(Closure {
+            fn_decl: FnDecl {
+                inputs: [Ty { hir_id: arg_id, .. }],
+                ..
+            },
+            body,
+            ..
+        }) => {
+            if let Node::Expr(expr) = cx.tcx.hir_node(body.hir_id) {
+                (peel_closure_body(cx, expr, *arg_id)).map(|body| (cx.tcx.hir().span(*arg_id), body))
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn peel_closure_body<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+    closure_arg_id: HirId,
+) -> Option<&'tcx Expr<'tcx>> {
+    match expr.kind {
+        // it would be nice if we could lift { x; y.a() } into { x; y }.a()
+        ExprKind::Block(
+            Block {
+                stmts: [],
+                expr: Some(wrapped_expr),
+                ..
+            },
+            _,
+        ) => peel_closure_body(cx, wrapped_expr, closure_arg_id),
+        ExprKind::MethodCall(_, pred, [arg], _) | ExprKind::Call(_, [pred, arg]) => {
+            if is_then_some(cx, expr) && is_local_defined_at(cx, arg, closure_arg_id) {
+                Some(pred)
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn is_local_defined_at(cx: &LateContext<'_>, local: &Expr<'_>, arg_hid: HirId) -> bool {
+    match local.kind {
+        ExprKind::Path(QPath::Resolved(
+            _,
+            Path {
+                res: Res::Local(local_hid),
+                ..
+            },
+        )) => {
+            // FIXME: this is the best way i could find to compare if
+            // a local refers to a specific closure argument.
+            //
+            // it breaks if the closure argument has an explicitly declared type,
+            // since the spans only align for TyKind::Infer
```

Comment:
Issue: You shouldn't use the `FnDecl` for this. The `Body` contains more parameter info, mostly importantly the `HirId`

See [Map::body](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.body), instead of getting the span of the body get the body itself :P

Reply:
the documentation on hir::Body seems to suggest it only contains argument patterns when used with a function, not a closure?

Reply:
It does for both cases. That documentation seems misleading tbh